src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                EventProcessed = new()
+                {
+                    Id = acknowledgement.MessageId,
+                    Status = new() { Status = acknowledgement.Action }
+                }
+            }, cancellationToken);
+        }
+    }
+
+    /// <summary>
+    /// Processes each topic messages from the channel as it's populated.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    private async Task ProcessTopicChannelMessagesAsync(CancellationToken cancellationToken)
+    {
+        await foreach (var message in topicMessagesChannel.Reader.ReadAllAsync(cancellationToken))
+        {
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            cts.CancelAfter(options.MessageHandlingPolicy.TimeoutDuration);
+
+            //Evaluate the message and return an acknowledgement result
+            var messageAction = await messageHandler(message, cts.Token);
+
+            try
+            {
+                //Share the result with the sidecar
+                await AcknowledgeMessageAsync(message.Id, messageAction);
+            }
+            catch (OperationCanceledException)
+            {
+                //Acknowledge the message using the configured default response action
+                await AcknowledgeMessageAsync(message.Id, options.MessageHandlingPolicy.DefaultResponseAction);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Retrieves the subscription stream data from the Dapr sidecar.
+    /// </summary>
+    /// <param name="stream">The stream connection to and from the Dapr sidecar instance.</param>
+    /// <param name="channelWriter">The channel writer instance.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    private async Task FetchDataFromSidecarAsync(
+        AsyncDuplexStreamingCall<P.SubscribeTopicEventsRequestAlpha1, P.SubscribeTopicEventsResponseAlpha1> stream,
+        ChannelWriter<TopicMessage> channelWriter, CancellationToken cancellationToken)
+    {
+        //Build out the initial topic events request
+        var initialRequest = new P.SubscribeTopicEventsRequestInitialAlpha1()
+        {
+            PubsubName = pubSubName, DeadLetterTopic = options.DeadLetterTopic ?? string.Empty, Topic = topicName
+        };
+
+        if (options.Metadata.Count > 0)
+        {
+            foreach (var (key, value) in options.Metadata)
+            {
+                initialRequest.Metadata.Add(key, value);
+            }
+        }
+        
+        //Send this request to the Dapr sidecar
+        await stream.RequestStream.WriteAsync(
+            new P.SubscribeTopicEventsRequestAlpha1 { InitialRequest = initialRequest }, cancellationToken);
+
+        //Each time a message is received from the stream, push it into the topic messages channel
+        await foreach (var response in stream.ResponseStream.ReadAllAsync(cancellationToken))
```

Comment:
I wonder if there will be a need to bound this (and perhaps make configurable)?  Forcing messages to stay "in the channel" would (should?) put backpressure on the Dapr runtime to keep it from potentially overwhelming the application or, in the case of multiple application replicas, prevent one replica from starving the other by accepting all the messages (but not yet processing them).  The Dapr runtime should manage this somewhat through round-robin-ing delivery amongst the connected streams, though.

Reply:
Trivial addition since we've got Channels under the hood. Added an optional configuration value to `DaprSubscriptionOptions` and if not null and > 0, it'll create a bounded channel instead of an unbounded channel and wait to push additional messages until there's room in the subscription loop.